Error when opening Rootcheck files on Windows

|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.2.0 | Rootcheck | Agent | All | Windows Server 2019 |

### Description

This is a bug reported by the @wazuh/qa team in the unstable version 4.2.

Sometimes, when the Windows agent is installed and started for the first time, its first Rootcheck scan is not able to find the benchmarks located at `$INSTALLDIR\ossec-agent\shared`, showing the following errors:

![image](https://user-images.githubusercontent.com/29475387/115577130-0cf70a00-a2bc-11eb-827e-a97af21da182.png)

The steps to reproduce it are:

- Install Wazuh 4.2.0 in a Windows Server 2019
- Connect it to a manager
- Wait until the first Rootcheck scan

Found an error that may be related to this one during the Windows installation test in a pull request:
``` 
================================== FAILURES ===================================
____________________________ test_check_log_errors ____________________________
install_agent_test.py:64: in test_check_log_errors
    assert expected_error == True, assert_description['error']['exception_not_detected'] + " in line {}".format(line)
E   AssertionError: An error was found in the log in line 2022/09/29 10:05:11 rootcheck: ERROR: (1252): Invalid rk configuration value: 'OSTS -> r'.
E     
E   assert False == True
E     -False
E     +True
```

--- a/src/client-agent/receiver.c
+++ b/src/client-agent/receiver.c
@@ -23,7 +23,6 @@
 static FILE *fp = NULL;
 static char file_sum[34] = "";
 static char file[OS_SIZE_1024 + 1] = "";
-static const char * IGNORE_LIST[] = { SHAREDCFG_FILENAME, NULL };
 #ifdef WIN32
 w_queue_t * winexec_queue;
 #endif
@@ -275,17 +274,19 @@ int receive_msg()
                         final_file = strrchr(file, '/');
                         if (final_file) {
                             if (strcmp(final_file + 1, SHAREDCFG_FILENAME) == 0) {
-                                if (cldir_ex_ignore(SHAREDCFG_DIR, IGNORE_LIST)) {
-                                    mwarn("Could not clean up shared directory.");
-                                }
-
-                                if(!UnmergeFiles(file, SHAREDCFG_DIR, OS_TEXT)){
+                                const char **IGNORE_LIST = malloc(2 * sizeof(char *));
+                                IGNORE_LIST[0] = SHAREDCFG_FILENAME;
+                                IGNORE_LIST[1] = NULL;
+                                if(!UnmergeFiles(file, SHAREDCFG_DIR, OS_TEXT, &IGNORE_LIST)){
                                     char msg_output[OS_MAXSTR];
 
                                     snprintf(msg_output, OS_MAXSTR, "%c:%s:%s",  LOCALFILE_MQ, "wazuh-agent", AG_IN_UNMERGE);
                                     send_msg(msg_output, -1);
                                 }
                                 else {
+                                    if (cldir_ex_ignore(SHAREDCFG_DIR, IGNORE_LIST)) {
+                                        mwarn("Could not clean up shared directory.");
+                                    }
                                     clear_merged_hash_cache();
                                     if (agt->flags.remote_conf && !verifyRemoteConf()) {
                                         if (agt->flags.auto_restart) {
@@ -296,6 +297,7 @@ int receive_msg()
                                         }
                                     }
                                 }
+                                os_free(IGNORE_LIST);
                             }
                         } else {
                             /* Remove file */

--- a/src/client-agent/receiver.c
+++ b/src/client-agent/receiver.c
@@ -274,10 +274,11 @@ int receive_msg()
                         final_file = strrchr(file, '/');
                         if (final_file) {
                             if (strcmp(final_file + 1, SHAREDCFG_FILENAME) == 0) {
-                                const char **IGNORE_LIST = malloc(2 * sizeof(char *));
-                                *IGNORE_LIST = strdup(SHAREDCFG_FILENAME);
+                                const char **IGNORE_LIST;
+                                os_malloc(2 * sizeof(char *), IGNORE_LIST);
+                                os_strdup(SHAREDCFG_FILENAME, *IGNORE_LIST);
                                 *(IGNORE_LIST + 1) = NULL;
-                                if(!UnmergeFiles(file, SHAREDCFG_DIR, OS_TEXT, &IGNORE_LIST)){
+                                if (!UnmergeFiles(file, SHAREDCFG_DIR, OS_TEXT, &IGNORE_LIST)) {
                                     char msg_output[OS_MAXSTR];
 
                                     snprintf(msg_output, OS_MAXSTR, "%c:%s:%s",  LOCALFILE_MQ, "wazuh-agent", AG_IN_UNMERGE);

--- a/src/client-agent/receiver.c
+++ b/src/client-agent/receiver.c
@@ -274,17 +274,17 @@ int receive_msg()
                         final_file = strrchr(file, '/');
                         if (final_file) {
                             if (strcmp(final_file + 1, SHAREDCFG_FILENAME) == 0) {
-                                const char **IGNORE_LIST;
-                                os_calloc(2, sizeof(char *), IGNORE_LIST);
-                                os_strdup(SHAREDCFG_FILENAME, *IGNORE_LIST);
-                                if (!UnmergeFiles(file, SHAREDCFG_DIR, OS_TEXT, &IGNORE_LIST)) {
+                                char **ignore_list;
+                                os_calloc(2, sizeof(char *), ignore_list);
+                                os_strdup(SHAREDCFG_FILENAME, *ignore_list);
+                                if (!UnmergeFiles(file, SHAREDCFG_DIR, OS_TEXT, &ignore_list)) {
                                     char msg_output[OS_MAXSTR];
 
                                     snprintf(msg_output, OS_MAXSTR, "%c:%s:%s",  LOCALFILE_MQ, "wazuh-agent", AG_IN_UNMERGE);
                                     send_msg(msg_output, -1);
                                 }
                                 else {
-                                    if (cldir_ex_ignore(SHAREDCFG_DIR, IGNORE_LIST)) {
+                                    if (cldir_ex_ignore(SHAREDCFG_DIR, ignore_list)) {
                                         mwarn("Could not clean up shared directory.");
                                     }
                                     clear_merged_hash_cache();
@@ -297,8 +297,7 @@ int receive_msg()
                                         }
                                     }
                                 }
-                                w_FreeArray(IGNORE_LIST);
-                                os_free(IGNORE_LIST);
+                                free_strarray(ignore_list);
                             }
                         } else {
                             /* Remove file */

--- a/src/client-agent/receiver.c
+++ b/src/client-agent/receiver.c
@@ -275,9 +275,8 @@ int receive_msg()
                         if (final_file) {
                             if (strcmp(final_file + 1, SHAREDCFG_FILENAME) == 0) {
                                 const char **IGNORE_LIST;
-                                os_malloc(2 * sizeof(char *), IGNORE_LIST);
+                                os_calloc(2, sizeof(char *), IGNORE_LIST);
                                 os_strdup(SHAREDCFG_FILENAME, *IGNORE_LIST);
-                                *(IGNORE_LIST + 1) = NULL;
                                 if (!UnmergeFiles(file, SHAREDCFG_DIR, OS_TEXT, &IGNORE_LIST)) {
                                     char msg_output[OS_MAXSTR];
 

--- a/src/client-agent/receiver.c
+++ b/src/client-agent/receiver.c
@@ -275,8 +275,8 @@ int receive_msg()
                         if (final_file) {
                             if (strcmp(final_file + 1, SHAREDCFG_FILENAME) == 0) {
                                 const char **IGNORE_LIST = malloc(2 * sizeof(char *));
-                                IGNORE_LIST[0] = SHAREDCFG_FILENAME;
-                                IGNORE_LIST[1] = NULL;
+                                *IGNORE_LIST = strdup(SHAREDCFG_FILENAME);
+                                *(IGNORE_LIST + 1) = NULL;
                                 if(!UnmergeFiles(file, SHAREDCFG_DIR, OS_TEXT, &IGNORE_LIST)){
                                     char msg_output[OS_MAXSTR];
 
@@ -296,6 +296,10 @@ int receive_msg()
                                             minfo("Shared agent configuration has been updated.");
                                         }
                                     }
+                                }
+                                int j = 0;
+                                while (*(IGNORE_LIST + j)) {
+                                    free(*(IGNORE_LIST + j++));
                                 }
                                 os_free(IGNORE_LIST);
                             }

--- a/src/client-agent/receiver.c
+++ b/src/client-agent/receiver.c
@@ -297,10 +297,7 @@ int receive_msg()
                                         }
                                     }
                                 }
-                                int j = 0;
-                                while (*(IGNORE_LIST + j)) {
-                                    free(*(IGNORE_LIST + j++));
-                                }
+                                w_FreeArray(IGNORE_LIST);
                                 os_free(IGNORE_LIST);
                             }
                         } else {

--- a/src/headers/file_op.h
+++ b/src/headers/file_op.h
@@ -218,7 +218,7 @@ int MergeAppendFile(FILE *finalfp, const char *files, int path_offset) __attribu
  * @param mode Indicates if the merged file must be readed as a binary file  or not. Use `#OS_TEXT`, `#OS_BINARY`.
  * @return 1 if the file was unmerged, 0 on error.
  */
-int UnmergeFiles(const char *finalpath, const char *optdir, int mode) __attribute__((nonnull(1)));
+int UnmergeFiles(const char *finalpath, const char *optdir, int mode, const char ***unmerged_files) __attribute__((nonnull(1)));
 
 
 /**

--- a/src/os_execd/wcom.c
+++ b/src/os_execd/wcom.c
@@ -119,7 +119,7 @@ size_t wcom_unmerge(const char *file_path, char ** output) {
         return strlen(*output);
     }
 
-    if (UnmergeFiles(final_path, INCOMING_DIR, OS_BINARY) == 0) {
+    if (UnmergeFiles(final_path, INCOMING_DIR, OS_BINARY, NULL) == 0) {
         merror("At WCOM unmerge: Error unmerging file '%s.'", final_path);
         os_strdup("err Cannot unmerge file", *output);
         return strlen(*output);

--- a/src/shared/file_op.c
+++ b/src/shared/file_op.c
@@ -768,7 +768,7 @@ int UnmergeFiles(const char *finalpath, const char *optdir, int mode, const char
             }
 
             /* Appends file name to unmerged files list */
-            *unmerged_files = realloc(*unmerged_files, (file_count + 1) * sizeof(char *));
+            *unmerged_files = realloc(*unmerged_files, (file_count + 2) * sizeof(char *));
             *(*unmerged_files + file_count) = strdup(file_name);
             file_count++;
         }

--- a/src/shared/file_op.c
+++ b/src/shared/file_op.c
@@ -644,7 +644,7 @@ int UnmergeFiles(const char *finalpath, const char *optdir, int mode, const char
     }
 
     /* Finds index of the last element on the list */
-    if(unmerged_files != NULL){
+    if (unmerged_files != NULL) {
         for(file_count = 0; *(*unmerged_files + file_count); file_count++);
     }
 
@@ -699,7 +699,6 @@ int UnmergeFiles(const char *finalpath, const char *optdir, int mode, const char
         }
 
         free(copy);
-
 
         /* Create temporary file */
         char tmp_file[strlen(final_name) + 7];
@@ -757,24 +756,24 @@ int UnmergeFiles(const char *finalpath, const char *optdir, int mode, const char
         /* Mv to original name */
         rename_ex(tmp_file, final_name);
 
-        if(unmerged_files != NULL){
+        if (unmerged_files != NULL) {
             /* Removes path from file name */
             file_name = strrchr(final_name, '/');
-            if(file_name){
+            if (file_name) {
                 file_name++;
             }
-            else{
+            else {
                 file_name = final_name;
             }
 
             /* Appends file name to unmerged files list */
-            *unmerged_files = realloc(*unmerged_files, (file_count + 2) * sizeof(char *));
-            *(*unmerged_files + file_count) = strdup(file_name);
+            os_realloc(*unmerged_files, (file_count + 2) * sizeof(char *), *unmerged_files);
+            os_strdup(file_name, *(*unmerged_files + file_count));
             file_count++;
         }
     }
 
-    if(unmerged_files != NULL){
+    if (unmerged_files != NULL) {
         *(*unmerged_files + file_count) = NULL;
     }
 

--- a/src/shared/file_op.c
+++ b/src/shared/file_op.c
@@ -623,15 +623,17 @@ void DeleteState() {
 }
 
 
-int UnmergeFiles(const char *finalpath, const char *optdir, int mode)
+int UnmergeFiles(const char *finalpath, const char *optdir, int mode, const char ***unmerged_files)
 {
     int ret = 1;
     int state_ok;
+    int file_count = 0;
     size_t i = 0, n = 0, files_size = 0;
     char *files;
     char * copy;
     char final_name[2048 + 1];
     char buf[2048 + 1];
+    char *file_name;
     FILE *fp;
     FILE *finalfp;
 
@@ -639,6 +641,11 @@ int UnmergeFiles(const char *finalpath, const char *optdir, int mode)
     if (!finalfp) {
         merror("Unable to read merged file: '%s' due to [(%d)-(%s)].", finalpath, errno, strerror(errno));
         return (0);
+    }
+
+    /* Finds index of the last element on the list */
+    if(unmerged_files != NULL){
+        for(file_count = 0; *(*unmerged_files + file_count); file_count++);
     }
 
     while (1) {
@@ -693,12 +700,22 @@ int UnmergeFiles(const char *finalpath, const char *optdir, int mode)
 
         free(copy);
 
+
+        /* Create temporary file */
+        char tmp_file[strlen(final_name) + 7];
+        snprintf(tmp_file, sizeof(tmp_file), "%sXXXXXX", final_name);
+
+        if (mkstemp_ex(tmp_file) == -1) {
+            merror("Unmerging '%s': could not create temporary file for '%s'", finalpath, files);
+            state_ok = 0;
+        }
+
         /* Open filename */
 
         if (state_ok) {
-            if (fp = fopen(final_name, mode == OS_BINARY ? "wb" : "w"), !fp) {
+            if (fp = fopen(tmp_file, mode == OS_BINARY ? "wb" : "w"), !fp) {
                 ret = 0;
-                merror("Unable to unmerge file '%s' due to [(%d)-(%s)].", final_name, errno, strerror(errno));
+                merror("Unable to unmerge file '%s' due to [(%d)-(%s)].", tmp_file, errno, strerror(errno));
             }
         } else {
             fp = NULL;
@@ -736,6 +753,29 @@ int UnmergeFiles(const char *finalpath, const char *optdir, int mode)
         if (fp) {
             fclose(fp);
         }
+
+        /* Mv to original name */
+        rename_ex(tmp_file, final_name);
+
+        if(unmerged_files != NULL){
+            /* Removes path from file name */
+            file_name = strrchr(final_name, '/');
+            if(file_name){
+                file_name++;
+            }
+            else{
+                file_name = final_name;
+            }
+
+            /* Appends file name to unmerged files list */
+            *unmerged_files = realloc(*unmerged_files, (file_count + 1) * sizeof(char *));
+            *(*unmerged_files + file_count) = strdup(file_name);
+            file_count++;
+        }
+    }
+
+    if(unmerged_files != NULL){
+        *(*unmerged_files + file_count) = NULL;
     }
 
     fclose(finalfp);

--- a/src/wazuh_modules/agent_upgrade/agent/wm_agent_upgrade_com.c
+++ b/src/wazuh_modules/agent_upgrade/agent/wm_agent_upgrade_com.c
@@ -353,7 +353,7 @@ STATIC char * wm_agent_upgrade_com_upgrade(const cJSON* json_object) {
     }
 
     //Unmerge
-    if (UnmergeFiles(merged, UPGRADE_DIR, OS_BINARY) == 0) {
+    if (UnmergeFiles(merged, UPGRADE_DIR, OS_BINARY, NULL) == 0) {
         unlink(merged);
         mterror(WM_AGENT_UPGRADE_LOGTAG, WM_UPGRADE_UNMERGING_FILE_ERROR, "upgrade", merged);
         return wm_agent_upgrade_command_ack(ERROR_UNMERGE, error_messages[ERROR_UNMERGE]);

